### PR TITLE
Fix bug in applying planning scene diffs that have attached collision objects

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

This fixes a bug we had while executing an MTC task where having two different planning scenes diffs that add a collision object to the scene, the first commit is just a test that trigger the bug, the second commit sets `robot_state.is_diff` to true when calling `getPlanningSceneDiffMsg` which prevent [conversions.cpp#L369-L372](https://github.com/ros-planning/moveit/blob/master/moveit_core/robot_state/src/conversions.cpp#L369-L372) from deleting the aco from the robot state when applying the diffs, but now when calling [PlanningScene::setCurrentState](https://github.com/ros-planning/moveit/blob/master/moveit_core/planning_scene/src/planning_scene.cpp#L1146) while detaching an object that object is staying in the scene so the last commit fixes it by explicitly removing the aco in the case of scene diff

This work is sponsored by RE2 Robotics

### CODE REVIEW

**Review by AndyZe — COMMENTED** [review]
I've tested this now with the Panda tutorial and clearing collision scene objects. It seems to work great! But, I still think you need to check if `parent_` is a nullptr.

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 1]
```diff
@@ -711,6 +711,18 @@ void PlanningScene::getPlanningSceneDiffMsg(moveit_msgs::PlanningScene& scene_ms
     if (do_omap)
       getOctomapMsg(scene_msg.world.octomap);
   }
+
+  // Ensure any detached collision objects get detached from the parent planning scene, too
+  for (const auto& collision_object : scene_msg.world.collision_objects)
```
Comment (simonschmeisser) [inline moveit_core/planning_scene/src/planning_scene.cpp:716]: What happens if the object is completely gone? Ie no longer attached but also removed from world.collision_objects?

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 2]
```diff
@@ -712,8 +712,7 @@ void PlanningScene::getPlanningSceneDiffMsg(moveit_msgs::PlanningScene& scene_ms
       getOctomapMsg(scene_msg.world.octomap);
   }
 
-  // This fixes an issue where detaching an object from the robot state diff is causing it to stay as attached collision
-  // object once applied back to the main scene, to get around this issue we make sure to delete it from robot state
+  // Ensure any detached collision objects get detached from the parent planning scene, too
```
Comment (JafarAbdi) [inline moveit_core/planning_scene/src/planning_scene.cpp:715]: Revert the change to this comment

File: moveit_core/planning_scene/src/planning_scene.cpp (hunk 1 above)
Comment (JafarAbdi) [reply moveit_core/planning_scene/src/planning_scene.cpp:716]: Do you mean after getting the diff as a message? I think in that case it shouldn't change the previous behavior

File: moveit_core/planning_scene/test/test_planning_scene.cpp [hunk 3]
```diff
@@ -345,6 +345,129 @@ TEST_P(CollisionDetectorTests, ClearDiff)
...
+
+// Returns collision objects names sorted alphabetically
+std::set<std::string> get_collision_objects_names(const planning_scene::PlanningScene& ps)
+{
+  std::vector<moveit_msgs::CollisionObject> collision_objects;
+  ps.getCollisionObjectMsgs(collision_objects);
+  std::set<std::string> collision_objects_names;
+  for (const auto& collision_object : collision_objects)
+    collision_objects_names.emplace(collision_object.id);
+  return collision_objects_names;
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_core/planning_scene/test/test_planning_scene.cpp]: I just remembered after applying the suggestion that `collision_objects_names` is a set of the IDs of the collision objects rather than the collision objects themselves, so this wouldn't compile :D

File: moveit_core/planning_scene/src/planning_scene.cpp [hunk 4]
```diff
@@ -711,6 +711,18 @@ void PlanningScene::getPlanningSceneDiffMsg(moveit_msgs::PlanningScene& scene_ms
     if (do_omap)
       getOctomapMsg(scene_msg.world.octomap);
   }
+
+  // Ensure any detached collision objects get detached from the parent planning scene, too
```
Comment (rhaschke) [inline moveit_core/planning_scene/src/planning_scene.cpp:715]: This comment is misleading: The parent planning scene is not (and should not get) modified.

**Review by rhaschke — COMMENTED** [review]
Sorry for being late on this PR. I have some more suggestions and comments. Please see #3142.

File: moveit_core/planning_scene/src/planning_scene.cpp (hunk 4 above)
Comment (JafarAbdi) [reply moveit_core/planning_scene/src/planning_scene.cpp:715]: Agreed, I had a to-do to update the comment, but it got resolved somehow
